### PR TITLE
Set `clearable` for Form `date` field when it is not required

### PR DESCRIPTION
**Pull request by andrearutrecht**
SVK-171

Set `clearable` for Form `date` field when it is not required.

### CODE REVIEW

**Comment by johnnyomair** [comment]
@andrearutrecht you probably want to target `next` for this, right?

**Comment by andrearutrecht** [comment]
`next` is the development default target branch, right? I should habe branched off of `next` then, too, instead of `main`. I will rebase.

**Comment by johnnyomair** [comment]
> `next` is the development default target branch, right? I should habe branched off of `next` then, too, instead of `main`. I will rebase.

No, next is the next major release (v7) and can include breaking changes. main is the current major version (v6). Changes that are non-breaking should typically target main.

You PR has many commits that are part of other branches. Since you're changing something in the future Admin Generator, you'll need to start from next. But please don't rebase, because this will result in duplicate commits.

**Comment by andrearutrecht** [comment]
I'll close this PR then and start a new one based on next.

**Comment by johnnyomair** [comment]
> I'll close this PR then and start a new one based on next?

Please create a local branch with the same name, add your changes, and force push the branch.

**Comment by johnnyomair** [comment]
@andrearutrecht please run the future Admin Generator and commit the changed files. This should fix the pipeline. Thanks!

**Comment by nsams** [comment]
Do we really want to clear button always when the field is required? (vs. making it configurable with maybe defaulting to true when required)

Why not add this feature to the field itself? (if not configurable)

**Comment by johnnyomair** [comment]
> Why not add this feature to the field itself? (if not configurable)

I don't think we want a clear button on every text input

> Do we really want to clear button always when the field is required? (vs. making it configurable with maybe defaulting to true when required)

Adding `clearable` to field config sounds like a good idea to me. The default could be `clearable={config.clearable ?? !config.required}`. This would allow setting `clearable: false` for some date fields.

**Comment by andrearutrecht** [comment]
@johnnyomair Is it your intention to add the default value for `clearable` to every `Field` type, or just for `date`?

**Comment by jamesricky** [comment]
> > @johnnyomair Is it your intention to add the default value for `clearable` to every `Field` type, or just for `date`?
> 
> Every field should support the `clearable` config option, but only date fields should use `!required` as default IMO

As discussed with the UX-Team (see SVK-171 > UX Meeting Notes) only date- and time-pickers should have have the clear-button, added by the `clearable` prop. It should always be there, when the field is not required, it should never be there when the field is not required. 

Selects, TextFields, etc. all have different methods of clearing the value, so `clearable` should not be supported by those fields. 

Since it's clearly defined, when the clear-button should be shown, configuration should not be needed.

**Comment by jamesricky** [comment]
> > As discussed with the UX-Team (see SVK-171 > UX Meeting Notes) only date- and time-pickers should have have the clear-button, added by the clearable prop. It should always be there, when the field is not required, it should never be there when the field is not required.
> 
> 👍
> 
> so, this could be a feature of the FinalFormDatePicker itself instead of admin-generator?

Good point, that would probably make sense 👍🏼 

@johnnyomair what do you think?

**Comment by jamesricky** [comment]
> > > > As discussed with the UX-Team (see SVK-171 > UX Meeting Notes) only date- and time-pickers should have have the clear-button, added by the clearable prop. It should always be there, when the field is not required, it should never be there when the field is not required.
> > > 
> > > 
> > > 👍
> > > so, this could be a feature of the FinalFormDatePicker itself instead of admin-generator?
> > 
> > 
> > Good point, that would probably make sense 👍🏼
> > @johnnyomair what do you think?
> 
> How would the FinalFormDataPicker know that it's required? Currently `required` is a feature of `Field`, not `FieldRenderProps`. However `DateField` etc. could do that.

I think it's just passed in with the other props. 
If `required` is missing, maybe we need to add it in the `Field`, like with the `disabled` prop in #1812.

**Comment by johnnyomair** [comment]
@andrearutrecht could you please test if this is possible doing in FinalFormDatePicker?

**Comment by andrearutrecht** [comment]
@johnnyomair I added the behaviour for `required` to `FinalFormDatePicker`. In consequence of that, all FinalFormFields now support  it. I tested all Fields listed in the file `AllFieldComponents`, none of which have issues when `required` is set.

The current implementation makes the prop `clearable` obsolete, as all Fields are now clearable by default. Should `clearable` be removed enitrely or is there a use case where it is still needed on its own?

**Comment by jamesricky** [comment]
> @johnnyomair I added the behaviour for `required` to `FinalFormDatePicker`. In consequence of that, all FinalFormFields now support it. I tested all Fields listed in the file `AllFieldComponents`, none of which have issues when `required` is set.
> 
> The current implementation makes the prop `clearable` obsolete, as all Fields are now clearable by default. Should `clearable` be removed enitrely or is there a use case where it is still needed on its own?

I don't see the need for the `clearable` prop if it's controlled by `required` now. 
Imo it can be removed.

**Comment by johnnyomair** [comment]
@andrearutrecht please remove the clearable prop altogether from the date picker, thanks!

File: .changeset/selfish-books-fail.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+"@comet/admin-date-time": major
+"@comet/admin": major
+---
+
+All `FinalFormField` components now support the `required` prop
+
+The `clearable` prop is set per default in `FinalFormDatePicker` when `required` is not set
```
Comment (jamesricky) [inline .changeset/selfish-books-fail.md:8]: > `clearable` prop is set per default

This isn't really correct anymore, as the prop has been removed.

Should this maybe be two separate changesets? 
- The "fix" for the `required` prop, that is now passed to all components
- The removal of the `clearable` prop and the clear-button being dependent on `required`

File: storybook/src/admin-date-time/DatePicker.tsx [hunk 2]
```diff
@@ -24,7 +26,8 @@ const Story = () => {
                         <Card>
                             <CardContent>
                                 <Field name="dateOne" label="Date" fullWidth component={FinalFormDatePicker} />
-                                <Field name="dateTwo" label="Clearable" fullWidth clearable component={FinalFormDatePicker} />
+                                <Field name="dateTwo" label="Required" fullWidth required component={FinalFormDatePicker} />
+                                <Field name="dateThree" label="Clearable" fullWidth component={FinalFormDatePicker} />
```
Comment (jamesricky) [inline storybook/src/admin-date-time/DatePicker.tsx:30]: I don't think `dateThree` is needed, as the non-required date above is already clearable now.

File: storybook/src/admin-date-time/DatePicker.tsx [hunk 3]
```diff
@@ -12,7 +12,7 @@ const Story = () => {
     }
 
     const initialValues: Partial<Values> = {
-        dateOne: null,
+        dateOne: new Date(),
```
Comment (johnnyomair) [inline storybook/src/admin-date-time/DatePicker.tsx:15]: Why is this necessary?

File: .changeset/selfish-books-fail.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
+"@comet/admin-date-time": major
+"@comet/admin": major
+---
+
+All `FinalFormField` components now support the `required` prop
+
+The `clearable` prop is removed from `FinalFormDatePicker`
+
+The clear button in `FinalFormDatePicker` is visible for all non-`required` fields
```
Comment (johnnyomair) [inline .changeset/selfish-books-fail.md:10]: Please split into two changesets, one for `@comet/admin` and one for `@comet/admin-date-time`. While the changes are related, they're still separate changes IMO.

File: storybook/src/admin-date-time/DatePicker.tsx (hunk 3 above)
Comment (andrearutrecht) [reply storybook/src/admin-date-time/DatePicker.tsx:15]: The clearable button is only visible when there is a Date set. That's why I changed this. I can add a third Date in to show the empty state. Should I do that?

File: .changeset/selfish-books-fail.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+"@comet/admin-date-time": major
+"@comet/admin": major
```
Comment (jamesricky) [inline .changeset/selfish-books-fail.md:3]: This is not specific to the `admin-date-time` package. 

```suggestion
"@comet/admin": major
```

File: .changeset/selfish-books-fail.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+---
+
+All `FinalFormField` components now support the `required` prop
```
Comment (johnnyomair) [inline .changeset/selfish-books-fail.md:5]:
```suggestion
Pass `required` prop to children of `Field` component
```

File: .changeset/selfish-books-fail.md (hunk 5 above)
Comment (johnnyomair) [reply .changeset/selfish-books-fail.md]: Also, I'd say this is a minor change.

File: .changeset/tidy-impalas-refuse.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+---
+"@comet/admin-date-time": major
+---
+
+The `clearable` prop is removed from `FinalFormDatePicker`
```
Comment (johnnyomair) [inline .changeset/tidy-impalas-refuse.md:5]:
```suggestion
Remove `clearable` prop from `FinalFormDatePicker`
```

File: .changeset/tidy-impalas-refuse.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+
+The clear button in `FinalFormDatePicker` is visible for all non-`required` fields
```
Comment (johnnyomair) [inline .changeset/tidy-impalas-refuse.md:7]:
```suggestion
The clear button will automatically be shown for all optional fields.
```

**Review by johnnyomair — APPROVED** [review]

**Comment by johnnyomair** [comment]
@andrearutrecht please fix conflicts.

**Comment by johnnyomair** [comment]
@jamesricky please review

File: .changeset/tidy-impalas-refuse.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+---
+"@comet/admin-date-time": major
+---
+
+Remove `clearable` prop from `FinalFormDatePicker`
```
Comment (jamesricky) [inline .changeset/tidy-impalas-refuse.md:5]: Technically it was removed from the `DatePicker`, which happens to affect the usage of `FinalFormDatePicker` and `DateField`. 

```suggestion
Remove `clearable` prop from `DatePicker`
```

or

```suggestion
Remove `clearable` prop from `DatePicker`, `DateField` and `FinalFormDatePicker`
```

**Review by jamesricky — CHANGES_REQUESTED** [review]

**Comment by jamesricky** [comment]
@thomasdax98 @johnnyomair for consistency, should we do this with all form components that currently have a `clearable` prop? Or at least the picker components? (In a separate PR)

- `ColorPicker`
- `DateRangePicker`
- `DateTimePicker`
- `TimePicker`
- `TimeRangePicker`
- `FinalFormSelect` (maybe)
- `Autocomplete` (maybe)
- `FinalFormInput` (maybe)
- `FinalFormSearchTextField` (maybe)

**Comment by thomasdax98** [comment]
@jamesricky 

I'd say yes for 

- ColorPicker
- DateRangePicker
- DateTimePicker
- TimePicker
- TimeRangePicker

But for I'd leave it for

- FinalFormSelect (maybe)
- Autocomplete (maybe)
- FinalFormInput (maybe)
- FinalFormSearchTextField (maybe)

because if you have a text input or multiple inputs (like a multi select) you might want to clear it even if it's required

**Comment by andrearutrecht** [comment]
@thomasdax98 Should the removal for the pickers happen in scope of this ticket or will this be a separate one?

**Comment by johnnyomair** [comment]
> @thomasdax98 Should the removal for the pickers happen in scope of this ticket or will this be a separate one?

@andrearutrecht could you please create a task for this?

Also, please fix conflicts, then this PR is good to go!

**Review by jamesricky — APPROVED** [review]

**Review by johnnyomair — APPROVED** [review]
